pixeltable/catalog/catalog.py

```diff
@@ -409,16 +412,45 @@ def begin_xact(
                         self._clear_tv_cache(handle.id, handle.effective_version)
                 self._modified_tvs.clear()
 
-    def raise_from_sql_exc(
-        self, e: sql_exc.StatementError, tbl_id: UUID | None, tbl: TableVersionHandle | None, convert_db_excs: bool
+    @contextmanager
+    def __begin_guarded_xact(self, *, for_write: bool) -> Iterator[sql.Connection]:
+        assert not self._current_tx_undo_actions
+        with Env.get().begin_xact(for_write=for_write) as conn:
+            try:
+                yield conn
+            except:
+                for hook in reversed(self._current_tx_undo_actions):
```

Comment:
Why not do this in begin_xact()?

Reply:
begin_xact() is getting awfully crowded. I started putting this in begin_xact() but to be correct it would have added a third level of try/finally nesting which I felt made begin_xact() almost unreadable. That's why I broke it out.

I'm of the opinion that begin_xact() should be broken down further into more digestible chunks (it's already quite hard to read) but I wasn't aiming to solve that problem here, just trying to avoid making things worse.